### PR TITLE
refactor(ingestion): establish setup for json consumption from kafka into clickhouse [nuke protobuf pt. 1]

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

See #9207 (this covers sections 1 and 2).

## Changes

* Create a new Kafka + MV table set for JSON encoded ingestion
* Update the sharding async migration to reflect this
* Stop using the protobuf topic on the Django server (only used for tests and demo data). 

I spent quite a bit of time trying to decide on the best way to update the migration (the current implementation wasn't very flexible) and ultimately landed on this approach.

## How did you test this code?

Tested manually + updated the existing tests.

### CODE REVIEW

File: ee/clickhouse/models/event.py [hunk 1]
```diff
@@ -42,19 +40,18 @@ def create_event(
...
+        "event": event,
+        "properties": json.dumps(properties),
+        "timestamp": timestamp.strftime("%Y-%m-%d %H:%M:%S.%f"),
+        "team_id": team.pk,
+        "distinct_id": str(distinct_id),
+        "elements_chain": elements_chain,
+        "created_at": timestamp.strftime("%Y-%m-%d %H:%M:%S.%f"),
+    }
+
+    ClickhouseProducer().produce(sql=INSERT_EVENT_SQL(), topic=KAFKA_EVENTS_JSON, data=kafka_event)
```
Comment (macobo) [inline ee/clickhouse/models/event.py:54]: Q: Is the protobuf path actually being executed anywhere in prod code?

File: ee/clickhouse/sql/events.py [hunk 2]
```diff
@@ -98,6 +98,35 @@
     database=settings.CLICKHOUSE_DATABASE,
 )
 
+KAFKA_EVENTS_TABLE_JSON_SQL = lambda: EVENTS_TABLE_BASE_SQL.format(
+    table_name="kafka_json_events",
```
Comment (macobo) [inline ee/clickhouse/sql/events.py:102]: Nit: WDYT about naming it `kafka_events` or `kafka_events_json`? Other tables are in form of `kafka_table`

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 3]
```diff
@@ -151,9 +152,11 @@ def replicated_table_operations(self, table: TableMigrationData):
             rollback=f"DROP TABLE IF EXISTS {table.tmp_table_name}",
         )
 
-        if table.kafka_table_name is not None:
+        for i in range(len(table.kafka_table_names)):
```
Comment (macobo) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:155]: `for create_kafka_table, kafka_table_name in zip(table.create_kafka_tables table.kafka_table_names)` 

Range is _often_ a sign that something has gone wrong

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 4]
```diff
@@ -173,22 +176,29 @@ def replicated_table_operations(self, table: TableMigrationData):
...
+            create_kafka_table = table.create_kafka_tables[i]
+            kafka_table_name = table.kafka_table_names[i]
             yield AsyncMigrationOperationSQL(
-                sql=cast(str, table.create_kafka_table), rollback=f"DROP TABLE IF EXISTS {table.kafka_table_name}"
+                sql=cast(str, create_kafka_table), rollback=f"DROP TABLE IF EXISTS {kafka_table_name}"
             )
 
-        if isinstance(table, ShardedTableMigrationData) and table.materialized_view_name is not None:
+        if isinstance(table, ShardedTableMigrationData):
+            for i in range(len(table.materialized_view_names)):
```
Comment (macobo) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:196]: zip

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 5]
```diff
@@ -173,22 +176,29 @@ def replicated_table_operations(self, table: TableMigrationData):
...
             for table_name, create_table_query in table.extra_tables.items():
                 yield AsyncMigrationOperationSQL(sql=create_table_query, rollback=f"DROP TABLE IF EXISTS {table_name}")
 
-        if isinstance(table, ShardedTableMigrationData) and table.materialized_view_name is not None:
-            yield AsyncMigrationOperationSQL(sql=f"DROP TABLE IF EXISTS {table.materialized_view_name}", rollback=None)
+            for mv_name in table.materialized_view_names:
+                yield AsyncMigrationOperationSQL(sql=f"DROP TABLE IF EXISTS {mv_name}", rollback=None)
 
-        if table.kafka_table_name is not None:
+        for i in range(len(table.kafka_table_names)):
```
Comment (macobo) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:188]: zip

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 6]
```diff
@@ -173,22 +176,29 @@ def replicated_table_operations(self, table: TableMigrationData):
...
+                sql=cast(str, create_kafka_table), rollback=f"DROP TABLE IF EXISTS {kafka_table_name}"
             )
 
-        if isinstance(table, ShardedTableMigrationData) and table.materialized_view_name is not None:
+        if isinstance(table, ShardedTableMigrationData):
+            for i in range(len(table.materialized_view_names)):
+                create_mv = table.create_materialized_views[i]
+                mv_name = table.materialized_view_names[i]
+
             yield AsyncMigrationOperationSQL(
```
Comment (macobo) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:200]: Bug: This has the wrong indentation. Only the last mat view is dropped.

File: ee/clickhouse/sql/events.py [hunk 7]
```diff
@@ -98,6 +98,35 @@
...
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=kafka_engine(topic=KAFKA_EVENTS_JSON),
```
Comment (macobo) [inline ee/clickhouse/sql/events.py:104]: The variable name and topic name don't agree - should we fix it? I guess not if others are similarly out of sync.

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 8]
```diff
@@ -173,22 +176,29 @@ def replicated_table_operations(self, table: TableMigrationData):
...
+            for mv_name in table.materialized_view_names:
+                yield AsyncMigrationOperationSQL(sql=f"DROP TABLE IF EXISTS {mv_name}", rollback=None)
 
-        if table.kafka_table_name is not None:
+        for i in range(len(table.kafka_table_names)):
+            create_kafka_table = table.create_kafka_tables[i]
+            kafka_table_name = table.kafka_table_names[i]
             yield AsyncMigrationOperationSQL(
-                sql=cast(str, table.create_kafka_table), rollback=f"DROP TABLE IF EXISTS {table.kafka_table_name}"
+                sql=cast(str, create_kafka_table), rollback=f"DROP TABLE IF EXISTS {kafka_table_name}"
```
Comment (macobo) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:192]: No need for the cast anymore.

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 9]
```diff
@@ -151,9 +152,11 @@ def replicated_table_operations(self, table: TableMigrationData):
             rollback=f"DROP TABLE IF EXISTS {table.tmp_table_name}",
         )
 
-        if table.kafka_table_name is not None:
+        for i in range(len(table.kafka_table_names)):
+            create_kafka_table = table.create_kafka_tables[i]
+            kafka_table_name = table.kafka_table_names[i]
             yield AsyncMigrationOperationSQL(
-                sql=f"DROP TABLE IF EXISTS {table.kafka_table_name}", rollback=cast(str, table.create_kafka_table)
+                sql=f"DROP TABLE IF EXISTS {kafka_table_name}", rollback=cast(str, create_kafka_table)
```
Comment (macobo) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:159]: No need for the cast anymore.

**Review by macobo — COMMENTED** [review]
Q: Should we just update plugin-server in a follow-up?

File: ee/clickhouse/models/event.py (hunk 1 above)
Comment (yakkomajuri) [reply ee/clickhouse/models/event.py:54]: nope, explained in the description

File: ee/clickhouse/sql/events.py (hunk 2 above)
Comment (yakkomajuri) [reply ee/clickhouse/sql/events.py:102]: sure

**Comment by yakkomajuri** [comment]
> Q: Should we just update plugin-server in a follow-up?

Plugin server should do things gradually. See #9215

File: posthog/async_migrations/migrations/0004_replicated_schema.py (hunk 6 above)
Comment (yakkomajuri) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:200]: good catch

File: ee/clickhouse/sql/events.py (hunk 7 above)
Comment (yakkomajuri) [reply ee/clickhouse/sql/events.py:104]: this follows the convention here

File: posthog/async_migrations/test/test_0004_replicated_schema.py [hunk 10]
```diff
@@ -94,8 +94,8 @@ def test_rollback(self):
         setup_async_migrations()
         migration = get_async_migration_definition(MIGRATION_NAME)
 
-        self.assertEqual(len(migration.operations), 53)
-        migration.operations[30].sql = "THIS WILL FAIL!"  # type: ignore
+        self.assertEqual(len(migration.operations), 56)
```
Comment (yakkomajuri) [inline posthog/async_migrations/test/test_0004_replicated_schema.py:97]:
```suggestion
        self.assertEqual(len(migration.operations), 57)
```

**Review by macobo — APPROVED** [review]
This looks good. Should we be super paranoid and just in case test out the sharding migration again on a freshly installed 1.33.0 install which then gets upgraded to a version with the new async migration?

**Comment by yakkomajuri** [comment]
@macobo wondering if we can punt on this - it won't make it into 1.34.0 so should be fine to get this in for Cloud.

Probably good to be paranoid and test though for 1.35.0

**Comment by macobo** [comment]
As long as we do it at some point I'm fine with it.

**Pull request merged**